Config deprecation updates (#8289)
fishtown-analytics/docs.getdbt.com: website/docs/reference/dbt-jinja-functions/config.md

1. The configuration variable exists, it is `None`
1. The configuration variable does not exist

:::warning Deprecation warning for meta fallback

> Starting in dbt Core v1.10, `config.get()` throws a deprecation warning when it finds a value in `config.meta` instead of at the top level. This fallback behavior will be removed in a future version.

Author: ```suggestion
Starting in <Constant name="core" /> v1.10, `config.get()` throws a deprecation warning when it finds a value in `config.meta` instead of at the top level. This fallback behavior will be removed in a future version.
```

> Starting in <Constant name="core" /> v1.10, `config.get()` throws a deprecation warning when it finds a value in `config.meta` instead of at the top level. This fallback behavior will be removed in a future version.

Reviewer: @gshank can comment on the specific dbt core version that will start warning

Reviewer 2: The dbt-protos and dbt-common parts of the change have been released. The dbt-core pull request will probably be merged tomorrow. It would go out in the "latest" cloud version probably next week?

Reviewer: spefically calling out, will it go in dbt core v1.10 or v1.11? Ik @jtcohen6 thought we may not need to backport to 1.10 and it may only end up in 1.11, is that right?

Author: Latest should align with 1.11 right now, correct? I move this up to 1.11 and if we ever backport, come back and change it

Reviewer 2: I think that sounds correct. We'll only backport to 1.10 if a product manager or somebody thinks it's a good idea.

Reviewer: I only called it out cause the highlighted text says 1.10!

> Starting in <Constant name="core" /> v1.11, `config.get()` throws a deprecation warning when it finds a value in `config.meta` instead of at the top level. This fallback behavior will be removed in a future version.

Reviewer 3: Trying to clarify the fallback step here so people don't think they can currently use top-level configs and just get a warning. 
```suggestion
Starting in <Constant name="core" /> v1.11, `config.get()` throws a deprecation warning when it finds a value in `config.meta`. This fallback was temporarily introduced when dbt reserved the top-level configs for official framework configuration. This fallback behavior will be removed in a future version.
```

Reviewer: agree!

  -- Example w/ default value. Default to 'id' if the 'unique_key' config does not exist
  {%- set unique_key = config.get('unique_key', default='id') -%}

>   -- Example of a custom config nested under `meta` as required in v1.10 and higher.

Reviewer: this is a valid alternative to config.meta_get() but i get how it could be confusing to show both

The `config.require` function is used to get configurations for a model from the end-user. Configs defined using this function are required, and failure to provide them will result in a compilation error.

> :::warning Deprecation warning for meta fallback

Reviewer 3: Should this be a reusable or are they different?

Author: There is a slight difference between them

> Starting in <Constant name=core" /> v1.10, `config.require()` throw a deprecation warning when it finds a value in `config.meta` instead of at the top level. This fallback behavior will be removed in a future version.

Author: ```suggestion
Starting in <Constant name="core" /> v1.10, `config.require()` throws a deprecation warning when it finds a value in `config.meta` instead of at the top level. This fallback behavior will be removed in a future version.
```

> Starting in <Constant name="core" /> v1.10, `config.require()` throws a deprecation warning when it finds a value in `config.meta` instead of at the top level. This fallback behavior will be removed in a future version.

Reviewer: same as above here on the core version

> Starting in <Constant name="core" /> v1.10, `config.require()` throws a deprecation warning when it finds a value in `config.meta` instead of at the top level. This fallback behavior will be removed in a future version.

Reviewer 3: Can we tell them what they are expected to do? Is it something like 
> Review the places you use custom configs under `meta` and either:
> * If the key belongs at the top level, move it out of meta.
> * If it intentionally lives in meta, update your code to config.meta_require().

Reviewer: the fusion error message does exactly this!
```bash
warning: dbt1000: DeprecationWarning: Custom config found under "meta" using config.require("old_required"). Please replace this with config.meta_require("old_required") to avoid collisions with configs introduced by dbt.
  --> models/test_require_deprecation.sql:9:35
warning: dbt1000: DeprecationWarning: Custom config found under "meta" using config.get("my_key"). Please replace this with config.meta_get("my_key") to avoid collisions with configs introduced by dbt.
  --> models/test_model.sql:8:25
  ```

Author: I can also link to the deprecation page which provides more context!

> Starting in dbt Core v1.10 and the <Constant name="fusion_engine" />, `config.require()` throw a deprecation warning when it finds a value in `config.meta` instead of at the top level. This fallback behavior will be removed in a future version.

Author: ```suggestion
Starting in <Constant name=core" /> v1.10, `config.require()` throw a deprecation warning when it finds a value in `config.meta` instead of at the top level. This fallback behavior will be removed in a future version.
```

To access custom configurations stored under `meta`, use [`config.meta_require()`](#configmeta_require) instead.
:::

Example usage:
```sql
{% materialization incremental, default -%}
  {%- set unique_key = config.require('unique_key') -%}
  ...
```

## config.meta_get

<VersionBlock lastVersion="1.9">

> This functionality is new in <Constant name="core" /> v1.10 and the <Constant name="fusion_engine" />.

Reviewer: same as above

> This functionality is new in dbt Core v1.10 and the <Constant name="fusion_engine" />.

Author: ```suggestion
This functionality is new in <Constant name="core" /> v1.10 and the <Constant name="fusion_engine" />.
```

</VersionBlock>

__Args__:

>  * `name`: The name of the configuration variable to retrieve from `meta` (required)

Author: ```suggestion
 - `name`: The name of the configuration variable to retrieve from `meta` (required)
```

>  * `default`: The default value to use if this configuration is not provided (optional)

Author: ```suggestion
 - `default`: The default value to use if this configuration is not provided (optional)
```

The `config.meta_get` function retrieves custom configurations stored under the `meta` dictionary. Unlike `config.get()`, this function exclusively checks `config.meta` and won't result in a deprecation warning.

> Use this function when accessing custom configurations that you've defined under `meta` in your model or resource configuration.

Reviewer: might be worth calling out that functionally its equivalent to writing config.get('meta').get(<key>)

Author: Will add the context!

Example usage:
```sql
{% materialization custom_materialization, default -%}
  -- Retrieve a custom config from meta, returns None if not found
  {%- set custom_setting = config.meta_get('custom_setting') -%}

  -- Retrieve with a default value
  {%- set custom_setting = config.meta_get('custom_setting', default='default_value') -%}
  ...
```

Example model configuration:
```yaml
models:
  - name: my_model
    config:
      meta:
        custom_setting: "my_value"
```

## config.meta_require

<VersionBlock lastVersion="1.9">

> This functionality is new in dbt Core v1.10 and the <Constant name="fusion_engine" />.

Author: ```suggestion
This functionality is new in <Constant name="core" /> v1.10 and the <Constant name="fusion_engine" />.
```

</VersionBlock>

__Args__:

>  * `name`: The name of the configuration variable to retrieve from `meta` (required)

Author: ```suggestion
 - `name`: The name of the configuration variable to retrieve from `meta` (required)
```

The `config.require` function is used to get configurations for a model from the end-user. Configs defined using this function are required, and failure to provide them will result in a compilation error.

:::warning Deprecation warning for meta fallback

> Starting in <Constant name="core" /> v1.11, `config.require()` throws a deprecation warning when it finds a value in `config.meta` instead of at the top level. This fallback behavior will be removed in a future version.

Reviewer 3: ```suggestion
Starting in <Constant name="core" /> v1.11, `config.require()` throws a deprecation warning when it finds a value in `config.meta`. This fallback was temporarily introduced when dbt reserved the top-level configs for official framework configuration. This fallback behavior will be removed in a future version.
```

Reviewer: same as above

:::

Example usage:
```sql
{% materialization incremental, default -%}
  {%- set unique_key = config.require('unique_key') -%}
  ...
```

## config.meta_get

<VersionBlock lastVersion="1.9">

This functionality is new in <Constant name="core" /> v1.11 and the <Constant name="fusion_engine" />.

</VersionBlock>

__Args__:

 - `name`: The name of the configuration variable to retrieve from `meta` (required)
 - `default`: The default value to use if this configuration is not provided (optional)

The `config.meta_get` function retrieves custom configurations stored under the `meta` dictionary. Unlike `config.get()`, this function exclusively checks `config.meta` and won't result in a deprecation warning.

Use this function when accessing custom configurations that you've defined under `meta` in your model or resource configuration - it's equivalent to writing `config.get('meta').get()`.

Example usage:
```sql
{% materialization custom_materialization, default -%}
  -- Retrieve a custom config from meta, returns None if not found
  {%- set custom_setting = config.meta_get('custom_setting') -%}

  -- Retrieve with a default value

>   {%- set custom_setting = config.meta_get('custom_setting', default='default_value') -%}

Reviewer 3: should this say `my_value` to match the yaml config?
```suggestion
  {%- set custom_setting = config.meta_get('my_value', default='default_value') -%}
```

Author: This was pulled directly from their ticket. I think this is calling the custom_setting name not the value. The value is set in the yaml.

Reviewer: or change both to custom_setting to clarify?

  ...
```

Example model configuration:
```yaml
models:
  - name: my_model
    config:
      meta:

>         custom_setting: "my_value"

Author: ```suggestion
        custom_setting: "custom_setting"
```

```

## config.meta_require

<VersionBlock lastVersion="1.9">

This functionality is new in <Constant name="core" /> v1.11 and the <Constant name="fusion_engine" />.

</VersionBlock>

__Args__:

 - `name`: The name of the configuration variable to retrieve from `meta` (required)

The `config.meta_require` function retrieves custom configurations stored under the `meta` dictionary. Unlike `config.require()`, this function exclusively checks `config.meta` and won't result in deprecation warnings. If the configuration is not found, dbt raises a compilation error.

Use this function when you need to ensure a custom configuration exists under `meta`.

Example usage:
```sql
{% materialization custom_materialization, default -%}
  -- Require a custom config from meta, throws error if not found
  {%- set required_setting = config.meta_require('required_setting') -%}
  ...

> ```

Reviewer 3: Should we show an example yaml config so they know how to set required setting in the YAML?

Reviewer: great suggestion!
